### PR TITLE
Feature/filter alicloud resources by tags

### PR DESCRIPTION
**Pull request by menglingwei**
filter alibabacloud resources (eg. SecurityGroupIDs ,VSwitchID) by tags

### CODE REVIEW

File: pkg/actuators/machine/instances.go [hunk 1]
```diff
@@ -363,68 +376,142 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
...
+		for _, sg := range machineProviderConfig.SecurityGroups {
+			if sg.ID != "" {
+				securityGroupIDs = append(securityGroupIDs, sg.ID)
+			} else {
+				if sg.Tags != nil {
+					request := ecs.CreateDescribeSecurityGroupsRequest()
+					if machineProviderConfig.VpcID != "" {
+						request.VpcId = machineProviderConfig.VpcID
+					}
+					if machineProviderConfig.ResourceGroupID != "" {
```
Comment (kwoodson) [inline pkg/actuators/machine/instances.go:397]: When `ResourceGroupID` is specified I am unable to query for the correct security group. I would remove resource group id.
```
aliyun ecs DescribeSecurityGroups --RegionId us-east-1 \
 --Tag.0.Key owned --Tag.0.Value "kubernetes.io/cluster/test-2mbh7" \
 --Tag.1.Key Name --Tag.1.Value "test-2mbh7-sg-worker" \
 --Tag.2.Key OCP --Tag.2.Value "ISV Integration"
{
	"PageNumber": 1,
	"PageSize": 10,
	"RegionId": "us-east-1",
	"RequestId": "BCD4DF49-FEED-57F6-A3AE-D670FB876026",
	"SecurityGroups": {
		"SecurityGroup": [
			{

				"CreationTime": "2021-11-01T17:11:52Z",
				"Description": "Created By OpenShift Installer",
				"ResourceGroupId": "",
				"SecurityGroupId": "sg-0xi9838n99r7i86s6rri",
				"SecurityGroupName": "test-2mbh7-sg-worker",
				"SecurityGroupType": "normal",
				"ServiceManaged": false,
				"Tags": {
					"Tag": [
						{
							"TagKey": "kubernetes.io/cluster/test-2mbh7",
							"TagValue": "owned"
						},
						{
							"TagKey": "Name",
							"TagValue": "test-2mbh7-sg-worker"
						},
						{
							"TagKey": "OCP",
							"TagValue": "ISV Integration"
						}
					]
				},
				"VpcId": "vpc-0xiosfgntngzl5ali1wj5"
...
```
Notice the `resourceGroupID` is not set.

File: pkg/actuators/machine/instances.go [hunk 2]
```diff
@@ -363,68 +376,142 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
...
+	case SecurityGroupTypeNoraml:
+		return MaxInstanceOfSecurityGroupTypeNoraml
+	case SecurityGroupTypeEnterprise:
+		return MaxInstanceOfSecurityGroupTypeEnterprise
+	default:
+		return MaxInstanceOfSecurityGroupTypeNoraml
 	}
+}
+
+func buildDescribeSecurityGroupsTag(tags []alibabacloudproviderv1.Tag) *[]ecs.DescribeSecurityGroupsTag {
```
Comment (kwoodson) [inline pkg/actuators/machine/instances.go:442]: This function returns a group of 6 security group tags. 3 empty ones and 3 correct ones which causes the query to fail.

Suggest fixing by using an index and storing them as you did with vswitch tags here:
https://github.com/openshift/cluster-api-provider-alibaba/pull/9/files#diff-c5079a7c5d2916c3661d0c44d1267050b3496c40b4530363d3825d66af581b6fR507-R508

File: pkg/actuators/machine/instances.go [hunk 3]
```diff
@@ -363,68 +376,142 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
...
+	describeSecurityGroupsTag := make([]ecs.DescribeSecurityGroupsTag, len(tags))
 
-	if len(response.SecurityGroups.SecurityGroup) < 1 {
-		klog.Errorf("no securitygroup for given filters not found")
-		return "", fmt.Errorf("no securitygroup for given filters not found")
+	for _, tag := range tags {
```
Comment (kwoodson) [inline pkg/actuators/machine/instances.go:445]: `for idx, tag := range tags {`

File: pkg/actuators/machine/instances.go [hunk 4]
```diff
@@ -363,68 +376,142 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
...
+		describeSecurityGroupsTag = append(describeSecurityGroupsTag, ecs.DescribeSecurityGroupsTag{
```
Comment (kwoodson) [inline pkg/actuators/machine/instances.go:446]: `describeSecurityGroupsTag[idx] = ecs.DescribeSecurityGroupsTag{`

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go [hunk 5]
```diff
@@ -127,74 +121,66 @@ type AlibabaCloudMachineProviderConfig struct {
...
-	//
 	//The default value is 40 or the size of the image, depending on whichever is greater.
 	SystemDiskSize int `json:"systemDiskSize,omitempty"`
 
 	//The description of the system disk. The description must be 2 to 256 characters in length and cannot start with http:// or https://.
-	//
 	//This parameter is empty by default.
 	SystemDiskDescription string `json:"systemDiskDescription,omitempty"`
 
 	DataDisks []DataDisk `json:"dataDisk,omitempty"`
```
Comment (kwoodson) [inline pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:136]: Add a d comment here.

**Comment by kwoodson** [comment]
@menglingwei I have left some feedback. We would like meet and chat about the cluster-api-provider-alibaba and the future plans of the provider types as well as these changes. Please sync up with Brian.

Thanks!

**Comment by kwoodson** [comment]
@menglingwei Also running into issues when accepting nodes into the cluster. We have an machine-approver that runs which accepts the node's certificate signing request. This is failing due to providerID being in a different format in two different locations. We will need to solve this as well so that nodes will be auto-accepted into the cluster.

cc @JoelSpeed @elmiko

File: pkg/actuators/machine/instances.go (hunk 4 above)
Comment (menglingwei) [reply pkg/actuators/machine/instances.go:446]: Let me fix it.

File: pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go (hunk 5 above)
Comment (menglingwei) [reply pkg/apis/alibabacloudprovider/v1beta1/alibabacloudmachineproviderconfig_types.go:136]: This line is redundant.

File: pkg/actuators/machine/instances.go (hunk 1 above)
Comment (menglingwei) [reply pkg/actuators/machine/instances.go:397]: I think if you don't assign an instance(ig. ECS Instance Or SecurityGroup ,etc) to a resource group. You don't need to set the resourceGroupID. If you set the resourceGroupID. The API simply returns the resources under a given resource group.

**Comment by menglingwei** [comment]
@kwoodson In Machine-API, we set the ProviderID follow the format  "alibabacloud:///RegionID/ZoneID/InstanceID". Do we need to keep the same ProviderID with CCM？

**Comment by kwoodson** [comment]
@menglingwei I think we should match what is in the CCM. The code that you reference is here:
https://github.com/openshift/cluster-api-provider-alibaba/blob/42ab2e529f632e4110930428fda16348a65c82e9/pkg/actuators/machine/reconciler.go#L312-L316

The CCM expects the code to be in this format `<region>.<instanceid>`. Here is the code in the CCM that parses the providerID https://github.com/openshift/cloud-provider-alibaba-cloud/blob/master/cloud-controller-manager/instances.go#L84.

I think we should match the CCM's expectation with `<region>.<instanceid>`.

@menglingwei WDYT?

**Comment by menglingwei** [comment]
> @menglingwei I think we should match what is in the CCM. The code in the cluster-api-provider-alibaba that you referenced is here:
> 
> https://github.com/openshift/cluster-api-provider-alibaba/blob/42ab2e529f632e4110930428fda16348a65c82e9/pkg/actuators/machine/reconciler.go#L312-L316
> 
> The CCM expects the code to be in this format `<region>.<instanceid>`. Here is the code in the CCM that parses the providerID https://github.com/openshift/cloud-provider-alibaba-cloud/blob/master/cloud-controller-manager/instances.go#L84.
> 
> I think we should match the CCM's expectation with `<region>.<instanceid>`.
> 
> @menglingwei WDYT?

It's ok.I'll do it.

File: pkg/actuators/machine/instances.go (hunk 1 above)
Comment (bd233) [reply pkg/actuators/machine/instances.go:397]: @kwoodson I think I can explain the problem, it should be that the resource_group is not specified [here](https://github.com/openshift/installer/blob/7a9bbc704b601a2c4af8821bfa3daeb8e9c64323/data/data/alibabacloud/cluster/vpc/sg-worker.tf#L1)
`resource_group_id` parameter, I have fixed it and submitted it in #5348

**Pull request merged**